torch/utils/checkpoint.py

```diff
@@ -335,7 +339,299 @@ def forward(input):
+# WeakKeyDictionary and pack strong references to the keys, so that as we
+# backward, those packed keys would be cleared as long as retain_graph=False.
+# Clearing the packed key clears the corresonding entry in the WKD.
+#
+# If we wish recomputed variables to be immediately cleared as we unpack them in
+# the retain_graph=True case, we cannot rely on the packed keys to be cleared by
+# backward automatically. Instead of packing the strong reference to the key
+# directly, we pack a container object, which we manually clear as we unpack.
+#
+# An important detail is that if a second backward happens, the second
+# recomputation needs to reset the container with a newly created key.
+#
+# Rule 5. Stop recomputation as soon as we've recomputed the saved tensors we
+#         know we need.
+#
+# [ Implementation details of Rule 5 ]
+#
+# During recomputation, raise an exception if the number of recomputed tensors
+# matches the number of tensors that we expected to recompute. We wrap the
+# recomputation call with a try-catch to catch this specific exception. See
+# Rule #5 below for some examples.
+#
+# Rule 6. We support doing backward inside checkpoint context
+#
+# This section is just a bunch of random examples that we'd like to support,
+# and comments on how that forced us to make certain design decisions.
+#
+# [ Basic case ]
+#
+# def fn(x):
+#   y = x.sin()
+#   z = y.cos()
+#   gx, = torch.autograd.grad(z, x, retains_grad=True)
+#   return gx, z
+#
+# out = checkpoint(fn)(inp)
+#
+# Because z is saved by cos while checkpoint is enabled, it would not be
+# actually saved, and so the .grad() call inside must trigger a recomputation.
+#
+# During recomputation the "inner pack hook" has two responsibilities:
+#
+# 1) As usual, populating the WeakKeyDictionary storing recomputed tensors
+# 2) Pack the tensor as-is so that one may perform backward on the recomputed
+#    graph. The tensors saved to this graph will live until the end of
+#    recomputation, or earlier if someone performs backward with
+#    retain_graph=False or something.
+#
+# [ Multiple backwards ]
+#
+# The example below shows what happens if during recomputation we find that some
+# of the tensors we are trying to recompute have already been cleared.
+#
+# Spoiler: we don't do anything special, we just skip over them!
+#
+# def fn(x):
+#   y = x.sin()                           # (1)
+#   z = y.cos()                           # (2)
+#   gx, = torch.autograd.grad(z, x)       # (3)
+#   w = x.sin()                           # (4)
+#   v = w.cos()                           # (5)
+#   gx2, = torch.autograd.grad(v, x)      # (6)
+#   return x * gx * gx2
+#
+# out = checkpoint(fn)(inp)
+#
+# In the code above fn is computed 4 times in total.
+#   1. Don't save x and y since we are inside a checkpoint.
+#   2. Trigger a recompute of fn as we reach (3) since x and y weren't saved.
+#   3. If early stop is enabled, stop at (2)
```

Comment:
Infinite recurse doesn't happen! (Check the tests :)
What happens is that during recompute when we run into backward again, this time it doesn't recompute again because this time we DO save things onto the graph.